src/azul/service/responseobjects/storage_service.py

```diff
@@ -39,12 +39,137 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
+            logger.error('Upload %s: Error detected within the MPU context.',
+                         self.upload_id,
+                         exc_info = (etype, value, traceback)
+                         )
+            self.__abort()
+        else:
+            self.__complete()
+
+    def __complete(self):
+        for future in as_completed(self.futures):
+            exception = future.exception()
+            if exception is not None:
+                logger.error('Upload %s: Error detected while uploading a part.',
+                             self.upload_id,
+                             exc_info=(type(exception), exception, None))
+                self.__abort()
+                raise UnexpectedMultipartUploadAbort(f'{self.bucket_name}/{self.object_key}') from exception
+
+        try:
+            self.mp_upload.complete(MultipartUpload={"Parts": [part.to_dict() for part in self.parts]})
+        except self.mp_upload.meta.client.exceptions.ClientError as exception:
+            logger.error('Upload %s: Error detected while completing the upload.',
+                         self.upload_id,
+                         exc_info=(type(exception), exception, None))
+            self.__abort()
+            if 'EntityTooSmall' in exception.args[0]:
+                raise UploadPartSizeOutOfBoundError(f'{self.bucket_name}/{self.object_key}')
+            else:
+                raise UnexpectedMultipartUploadAbort(f'{self.bucket_name}/{self.object_key}')
+
+        self.mp_upload = None
+        self.thread_pool.shutdown()
+
+    def __abort(self):
+        logger.info('Upload %s: Aborting', self.upload_id)
+        # This implementation will ignore any pending/active part uploads and force the thread pool to shut down.
+        self.mp_upload.abort()
+        self.mp_upload = None
+        self.thread_pool.shutdown(wait=False)
+        logger.warning('Upload %s: Aborted', self.upload_id)
+
+    def push(self, data: bytes):
+        part = self._create_new_part(data)
+        self.futures.append(self.thread_pool.submit(self._upload_part, part))
+
+    def _create_new_part(self, data: bytes):
+        part = Part(part_number=self.next_part_number, etag=None, content=data)
+        self.parts.append(part)
+        self.next_part_number += 1
+
+        return part
+
+    def _upload_part(self, part):
+        upload_part = self.mp_upload.Part(part.part_number)
+        result = upload_part.upload(Body=part.content)
+        part.etag = result['ETag']
+
+
+@dataclass
+class Part:
+    etag: str  # If ETag is defined, the content is already pushed to S3.
+    part_number: int
+    content: bytes
+
+    @property
+    def already_uploaded(self):
+        return self.etag is not None
+
+    def to_dict(self):
+        return dict(PartNumber=self.part_number, ETag=self.etag)
 
 
 class GetObjectError(RuntimeError):
     pass
+
+
+class EmptyMultipartUploadError(RuntimeError):
+    pass
+
+
+class UnexpectedMultipartUploadAbort(RuntimeError):
```

Comment:
Give this class a constructor that requires bucket name and key as arguments and compose the message to the super constructor. That way you can avoid having to compose the message `f'{self.bucket_name}/{self.object_key}` at each call site.

Reply:
```suggestion
class MultipartUploadError(RuntimeError):
```